File: core/baker-interface-kotlin/src/main/kotlin/com/ing/baker/runtime/kotlindsl/RecipeInstanceConfig.kt
package com.ing.baker.runtime.kotlindsl

import com.ing.baker.runtime.model.recipeinstance.RecipeInstance
import scala.Option
import scala.concurrent.duration.FiniteDuration
import scala.jdk.CollectionConverters
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

data class RecipeInstanceConfig(
        val idleTTL: Duration? = 5.seconds,
        val ingredientsFilter: List<String> = emptyList()) {

    private fun Duration.toFiniteDuration(): FiniteDuration {
        return FiniteDuration.fromNanos(this.inWholeNanoseconds)
    }
    fun toBakerFRecipeInstanceConfig(): RecipeInstance.Config {
        return RecipeInstance.Config(
                idleTTL?.let{ Option.apply(it.toFiniteDuration()) } ?: Option.empty(),
                CollectionConverters.ListHasAsScala(ingredientsFilter).asScala().toSeq()
        )
    }
}

File: core/baker-interface/src/main/scala/com/ing/baker/runtime/javadsl/RecipeInstanceConfig.scala
package com.ing.baker.runtime.javadsl

import com.ing.baker.runtime.model.recipeinstance.RecipeInstance

import java.time.Duration
import java.util.Optional
import scala.concurrent.duration.{FiniteDuration, NANOSECONDS}
import scala.jdk.CollectionConverters.CollectionHasAsScala
import scala.jdk.OptionConverters.RichOptional

case class RecipeInstanceConfig(idleTTL: Optional[Duration] = Optional.of(Duration.ofSeconds(5)),
                                ingredientsFilter: java.util.List[String] = java.util.List.of()) {

  def withIdleTTL(idleTTL: Optional[Duration]) = copy(idleTTL = idleTTL)

  def withIngredientsFilter(ingredientsFilter: java.util.List[String]) = ingredientsFilter

  def toBakerFRecipeInstanceConfig(): RecipeInstance.Config = {
    RecipeInstance.Config(
      idleTTL.toScala.map(duration => FiniteDuration.apply(duration.toNanos, NANOSECONDS)),
      ingredientsFilter.asScala.toSeq
    )
  }
}
